README.md
# libimengines

debian/changelog
libimengines (4.2.5) unstable; urgency=low

  * Fix imengines_wp_word_exists signature

 -- Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>  Sun, 23 Feb 2020 18:32:13 +0200

libimengines (4.2.4) unstable; urgency=low

  * Initial Maemo Leste import

 -- Merlijn Wajer <merlijn@wizzup.org>  Mon, 9 Jul 2018 23:36:00 +0200

libimengines (4.2.3) unstable; urgency=low

  * Fix packaging

 -- Jonathan Wilson <jfwfreo@tpgi.com.au>  Sat, 28 Oct 2017 21:26:57 -0400

libimengines (4.2.2) unstable; urgency=low

  * Initial RE release.

 -- Jonathan Wilson <jfwfreo@tpgi.com.au>  Sat, 28 Oct 2017 20:02:58 -0400


imengines.c
#include <string.h>

static const struct
{
  char locale[6];
}
locales[]=
{
  { "af_ZA" },
  { "am_ET" },
  { "ar_AE" },
  { "ar_BH" },
  { "ar_DZ" },
  { "ar_EG" },
  { "ar_IN" },
  { "ar_IQ" },
  { "ar_JO" },
  { "ar_KW" },
  { "ar_LB" },
  { "ar_LY" },
  { "ar_MA" },
  { "ar_OM" },
  { "ar_QA" },
  { "ar_SA" },
  { "ar_SD" },
  { "ar_SY" },
  { "ar_TN" },
  { "ar_YE" },
  { "az_AZ" },
  { "be_BY" },
  { "bg_BG" },
  { "bn_IN" },
  { "br_FR" },
  { "bs_BA" },
  { "ca_ES" },
  { "cs_CZ" },
  { "cy_GB" },
  { "da_DK" },
  { "de_AT" },
  { "de_BE" },
  { "de_CH" },
  { "de_DE" },
  { "de_LU" },
  { "el_GR" },
  { "en_AU" },
  { "en_BW" },
  { "en_CA" },
  { "en_DK" },
  { "en_GB" },
  { "en_HK" },
  { "en_IE" },
  { "en_IN" },
  { "en_NZ" },
  { "en_PH" },
  { "en_SG" },
  { "en_US" },
  { "en_ZA" },
  { "en_ZW" },
  { "eo_EO" },
  { "es_AR" },
  { "es_BO" },
  { "es_CL" },
  { "es_CO" },
  { "es_CR" },
  { "es_DO" },
  { "es_EC" },
  { "es_ES" },
  { "es_GT" },
  { "es_HN" },
  { "es_MX" },
  { "es_NI" },
  { "es_PA" },
  { "es_PE" },
  { "es_PR" },
  { "es_PY" },
  { "es_SV" },
  { "es_US" },
  { "es_UY" },
  { "es_VE" },
  { "et_EE" },
  { "eu_ES" },
  { "fa_IR" },
  { "fi_FI" },
  { "fo_FO" },
  { "fr_BE" },
  { "fr_CA" },
  { "fr_CH" },
  { "fr_FR" },
  { "fr_LU" },
  { "ga_IE" },
  { "gd_GB" },
  { "gl_ES" },
  { "gv_GB" },
  { "he_IL" },
  { "hi_IN" },
  { "hr_HR" },
  { "hu_HU" },
  { "hy_AM" },
  { "id_ID" },
  { "is_IS" },
  { "it_CH" },
  { "it_IT" },
  { "iw_IL" },
  { "ja_JP" },
  { "ka_GE" },
  { "kl_GL" },
  { "ko_KR" },
  { "kw_GB" },
  { "lt_LT" },
  { "lv_LV" },
  { "mi_NZ" },
  { "mk_MK" },
  { "mr_IN" },
  { "ms_MY" },
  { "mt_MT" },
  { "nl_BE" },
  { "nl_NL" },
  { "nn_NO" },
  { "no_NO" },
  { "oc_FR" },
  { "pl_PL" },
  { "pt_BR" },
  { "pt_PT" },
  { "ro_RO" },
  { "ru_RU" },
  { "ru_UA" },
  { "se_NO" },
  { "sk_SK" },
  { "sl_SI" },
  { "sq_AL" },
  { "sr_YU" },
  { "sv_FI" },
  { "sv_SE" },
  { "ta_IN" },
  { "te_IN" },
  { "tg_TJ" },
  { "th_TH" },
  { "ti_ER" },
  { "ti_ET" },
  { "tl_PH" },
  { "tr_TR" },
  { "tt_RU" },
  { "uk_UA" },
  { "ur_PK" },
  { "uz_UZ" },
  { "vi_VN" },
  { "wa_BE" },
  { "yi_US" },
  { "zh_CN" },
  { "zh_HK" },
  { "zh_SG" },
  { "zh_TW" }
};

#define NUM_LOCALES (sizeof(locales) / sizeof(locales[0]))

const char *
imengines_locale_to_string(int locale)
{
  if (locale < NUM_LOCALES)
    return locales[locale].locale;

  return NULL;
}

int
imengines_string_to_locale(const char *s)
{
  int i;

  if (!s)
    return -1;

  for (i = 0; i < NUM_LOCALES; i++)
  {
    if (!strncmp(locales[i].locale, s, sizeof(locales[0]) - 1))
      return i;
  }

  return -1;
}


imengines-wp.h
/**
  @file imengines-wp.h

  Copyright (C) 2017 Jonathan Wilson

  @author Jonathan wilson <jfwfreo@tpgi.com.au>

  This library is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License version 2.1 as
  published by the Free Software Foundation.

  This library is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this library; if not, write to the Free Software Foundation,
  Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

typedef struct _imengines_wp_candidates {
  int number_of_candidates;
  char candidate[10][128];
} imengines_wp_candidates;

typedef enum _imengines_im_status {
  IM_STATUS_UNKNOWN = -1,
  IM_STATUS_OK = 0,
  IM_STATUS_FAILED,
  IM_STATUS_OUT_OF_RANGE,
  IM_STATUS_NOT_ENOUGHT_MEMORY,
  IM_STATUS_OUT_OF_LIMITS,
  IM_STATUS_OUT_OF_DISKSPACE,
  IM_STATUS_NOT_SUPPORTED
} imengines_im_status;
typedef struct _imengines_wp_plugin {
  void (*destroy)(void);
  imengines_im_status (*set_language)(const char *language, unsigned int language_index);
  imengines_im_status (*set_data)(const char *d, const char *v);
  imengines_im_status (*attach_dictionary)(unsigned int dictionary, gboolean use_dictionary);
  imengines_im_status (*detach_dictionary)(unsigned int dictionary);
  imengines_im_status (*get_candidates)(const gchar *previous_word, const gchar *current_word, imengines_wp_candidates *candidates);
  imengines_im_status (*set_max_candidates)(int max_candidates);
  imengines_im_status (*get_max_candidates)(int *number);
  gboolean (*word_exists)(const char *word, unsigned int dictionary, unsigned int *language_index);
  imengines_im_status (*hit_word)(const char *word, unsigned int dictionary, unsigned int language_index);
  imengines_im_status (*delete_word)(const char *word, unsigned int dictionary, unsigned int language_index);
  imengines_im_status (*add_word)(const char *word, unsigned int dictionary, unsigned int language_index);
  imengines_im_status (*save_dictionary)(unsigned int dictionary);
  int (*get_number_supported_languages)(void);
} imengines_wp_plugin;
int imengines_wp_get_number_supported_languages(void);
imengines_im_status imengines_wp_save_dictionary(unsigned int dictionary);
imengines_im_status imengines_wp_add_word(const char *word, unsigned int dictionary, unsigned int language_index);
imengines_im_status imengines_wp_delete_word(const char *word, unsigned int dictionary, unsigned int language_index);
imengines_im_status imengines_wp_hit_word(const char *word, unsigned int dictionary, unsigned int language_index);
gboolean imengines_wp_word_exists(const char *word, unsigned int dictionary, unsigned int *language_index);
imengines_im_status imengines_wp_get_max_candidates(int *number);
imengines_im_status imengines_wp_set_max_candidates(int max_candidates);
imengines_im_status imengines_wp_get_candidates(const gchar *previous_word, const gchar *current_word, imengines_wp_candidates *candidates);
imengines_im_status imengines_wp_detach_dictionary(unsigned int dictionary);
imengines_im_status imengines_wp_attach_dictionary(unsigned int dictionary, gboolean use_dictionary);
imengines_im_status imengines_wp_set_data(const char *d, const char *v);
imengines_im_status imengines_wp_set_prediction_language(const char *language, unsigned int language_index);
void imengines_wp_destroy(void);
imengines_im_status imengines_wp_init(const char *wpengine);


imengines-wp.c
/**
  @file imengines-wp.c

  Copyright (C) 2017 Jonathan Wilson

  @author Jonathan wilson <jfwfreo@tpgi.com.au>

  This library is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License version 2.1 as
  published by the Free Software Foundation.

  This library is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this library; if not, write to the Free Software Foundation,
  Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/
#include <glib.h>
#include <stdio.h>
#include <string.h>
#include <dlfcn.h>
#include "imengines-wp.h"

static imengines_wp_plugin *wp;
static void *wp_handle;

int
imengines_wp_get_number_supported_languages(void)
{
  g_return_val_if_fail(wp != NULL, 1);
  g_return_val_if_fail(wp->get_number_supported_languages != NULL, 1);

  return wp->get_number_supported_languages();
}

imengines_im_status
imengines_wp_save_dictionary(unsigned int dictionary)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->save_dictionary != NULL, IM_STATUS_FAILED);

  return wp->save_dictionary(dictionary);
}

imengines_im_status
imengines_wp_add_word(const char *word, unsigned int dictionary,
                      unsigned int language_index)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->add_word != NULL, IM_STATUS_FAILED);

  return wp->add_word(word, dictionary, language_index);
}

imengines_im_status
imengines_wp_delete_word(const char *word, unsigned int dictionary,
                         unsigned int language_index)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->delete_word != NULL, IM_STATUS_FAILED);

  return wp->delete_word(word, dictionary, language_index);
}

imengines_im_status
imengines_wp_hit_word(const char *word, unsigned int dictionary,
                      unsigned int language_index)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->hit_word != NULL, IM_STATUS_FAILED);

  return wp->hit_word(word, dictionary, language_index);
}

gboolean
imengines_wp_word_exists(const char *word, unsigned int dictionary,
                         unsigned int *language_index)
{
  g_return_val_if_fail(wp != NULL,FALSE);
  g_return_val_if_fail(wp->word_exists != NULL, FALSE);

  return wp->word_exists(word, dictionary, language_index);
}

imengines_im_status
imengines_wp_get_max_candidates(int *number)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->get_max_candidates != NULL, IM_STATUS_FAILED);

  return wp->get_max_candidates(number);
}

imengines_im_status
imengines_wp_set_max_candidates(int max_candidates)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->set_max_candidates != NULL, IM_STATUS_FAILED);

  return wp->set_max_candidates(max_candidates);
}

imengines_im_status
imengines_wp_get_candidates(const gchar *previous_word,
                            const gchar *current_word,
                            imengines_wp_candidates *candidates)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->get_candidates != NULL, IM_STATUS_FAILED);

  return wp->get_candidates(previous_word, current_word, candidates);
}

imengines_im_status
imengines_wp_detach_dictionary(unsigned int dictionary)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->detach_dictionary != NULL, IM_STATUS_FAILED);

  return wp->detach_dictionary(dictionary);
}
imengines_im_status
imengines_wp_attach_dictionary(unsigned int dictionary, gboolean use_dictionary)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->attach_dictionary != NULL, IM_STATUS_FAILED);

  return wp->attach_dictionary(dictionary, use_dictionary);
}

imengines_im_status imengines_wp_set_data(const char *d, const char *v)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->set_data != NULL, IM_STATUS_FAILED);

  return wp->set_data(d, v);
}

imengines_im_status
imengines_wp_set_prediction_language(const char *language,
                                     unsigned int language_index)
{
  g_return_val_if_fail(wp != NULL, IM_STATUS_FAILED);
  g_return_val_if_fail(wp->set_language != NULL, IM_STATUS_FAILED);

  return wp->set_language(language, language_index);
}

void
imengines_wp_destroy(void)
{
  g_return_if_fail(wp != NULL);
  g_return_if_fail(wp->destroy != NULL);

  wp->destroy();
  wp = 0;
}

imengines_im_status
imengines_wp_init(const char *wpengine)
{
  char str[1024];
  imengines_wp_plugin *(*func)();
  memset(str, 0, sizeof(str));
  /* FIXME - multiarch */
  snprintf(str, sizeof(str) - 1, "%s/libim_wp_plugin_%s.so",
           "/usr/lib/libimengines", wpengine);

  if (wp)
    imengines_wp_destroy();

  wp = NULL;
  wp_handle = dlopen(str, RTLD_NOW);

  if (wp_handle)
  {
    func = (imengines_wp_plugin *(*)())dlsym(wp_handle,
                                             "imengines_wp_plugin_init");

    if (func)
    {
      wp = func();

      if (wp)
        return IM_STATUS_OK;

      g_warning("Failed calling init function for WP");
    }
    else
      g_warning("Can't find init function in WP plugin %s", wpengine);
  }
  else
    g_warning( "Can't open WP plugin library %s", str);

  if (wp_handle)
    dlclose(wp_handle);

  return IM_STATUS_FAILED;
}
